server/auth_callout.go

```diff
@@ -0,0 +1,329 @@
+	respCh := make(chan bool, 1)
+
+	processReply := func(_ *subscription, rc *client, racc *Account, subject, reply string, rmsg []byte) {
+		_, msg := rc.msgParts(rmsg)
+		// This signals not authorized.
+		// Since this is an account subscription will always have "\r\n".
+		if len(msg) == LEN_CR_LF {
+			respCh <- false
+			return
+		}
+		// Strip trailing CRLF.
+		msg = msg[:len(msg)-LEN_CR_LF]
+
+		// If we sent an encrypted request the response could be encrypted as well.
+		if xkp != nil && len(msg) > len(jwtPrefix) && !bytes.HasPrefix(msg, []byte(jwtPrefix)) {
+			var err error
+			msg, err = xkp.Open(msg, pubAccXKey)
+			if err != nil {
+				s.Warnf("Error decrypting auth callout response: %v", err)
+				respCh <- false
+				return
+			}
+		}
+
+		juc, err := jwt.DecodeUserClaims(string(msg))
+		if err != nil {
+			s.Warnf("Error decoding auth callout response: %v")
+			respCh <- false
+			return
+		}
+
+		// Make sure that the user is what we requested.
+		if juc.Subject != pub {
+			s.Warnf("Expected authorized user of %q but got %q", pub, juc.Subject)
+			respCh <- false
+			return
+		}
+
+		// Make sure correct issuer.
+		var issuer string
+		if opts.AuthCallout != nil {
+			issuer = opts.AuthCallout.Issuer
+		} else {
+			// Operator mode is who we send the request on unless switching accounts.
+			issuer = acc.Name
+		}
+		// By default issuer needs to match server config or the requesting account in operator mode.
+		if juc.Issuer != issuer {
+			if !isOperatorMode {
+				s.Warnf("Wrong issuer for auth callout response: %v")
+				respCh <- false
+				return
+			} else if !acc.isAllowedAcount(juc.Issuer) {
+				s.Warnf("Account %q not allowed for auth callout on %q", juc.Issuer, issuer)
+				respCh <- false
+				return
+			}
+		}
+		allowNow, validFor := validateTimes(juc)
+		if !allowNow {
+			c.Errorf("Outside connect times")
+			respCh <- false
+			return
+		}
+		allowedConnTypes, err := convertAllowedConnectionTypes(juc.AllowedConnectionTypes)
+		if err != nil && len(allowedConnTypes) == 0 {
+			respCh <- false
+			return
+		}
+		// Apply to this client.
+		targetAcc := acc
+		// Check if we are being asked to switch accounts.
+		if aname := juc.Audience; aname != _EMPTY_ {
+			targetAcc, err = s.LookupAccount(aname)
+			if err != nil {
+				s.Warnf("No valid account %q for auth callout response: %v", aname, err)
+				respCh <- false
+				return
+			}
+			// In operator mode make sure this account matches the issuer.
+			if isOperatorMode && aname != juc.Issuer {
+				s.Warnf("Account %q does not match issuer %q", aname, juc.Issuer)
+				respCh <- false
+				return
+			}
+		}
+
+		// Build internal user and bind to the targeted account.
+		nkuser := buildInternalNkeyUser(juc, allowedConnTypes, targetAcc)
+		if err := c.RegisterNkeyUser(nkuser); err != nil {
+			s.Warnf("Could not register auth callout user: %v", err)
+			respCh <- false
+			return
+		}
+
+		// See if the response wants to override the userID.
```

Comment:
This comment doesn't appear to describe the actions to be taken

Reply:
Will fix.